rootfs/etc/nginx/template/nginx.tmpl

```diff
@@ -952,14 +952,33 @@ stream {
 
         set $proxy_upstream_name "-";
 
+        {{ if not ( empty $server.CertificateAuth.ExactMatchCN ) }}
```

Comment:
We don't add a XOR condition in the annotation parser, saying a user can configure ExactMatchCN OR MatchCN.

This leads me to think this will work, per the template like
* User sets an exactMatchCN
* User then sets a MatchCN
* per Template run, matchCN will replace exactMatchCN

As this is not documented, I would expect that we define a behavior here and use this behavior in the parser, and not in the template. Let me suggest that:

* In the parser, we verify first one annotation, then another. As ExactMatchCN is more restrictive (and new annotation) i would suggest that ExactMatchCN takes precedence over MatchCN, so someone configuring it will know what they are doing.
* We can add in validation webhook a denial saying that this annotations are mutually exclusive if someone tries to add them both. Same on the parser, if both are configured, return an error.
* If you look into the template code, you are going to see that you are repeating the ifs with almost the same information but the name in the struct (ExactMatchCN) and the operation `if ( $ssl_client_s_dn_cn != {{ $server.CertificateAuth.ExactMatchCN }} )` vs `if ( $ssl_client_s_dn_cn !~ {{ $server.CertificateAuth.MatchCN }} ) {`

So, what if you instead of creating a new string annotation, creates a new bool annotation that says if CN should exactly match, and based on this (if true) you change the way your annotation works/get parsed, and the way the template works?

I think this will be the least impact possible on the way it currently works. WDYT?

Reply:
This makes sense and is something I overlooked. I definitely don't want people using both and having matchCN replace exactMatchCN. I love this idea of adding in the bool annotation, this will work very well I believe!